executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,686 @@
+	dLookUpKeys, err := imw.constructDatumLookupKeys(task)
+	if err != nil {
+		return err
+	}
+	dLookUpKeys = imw.dedupDatumLookUpKeys(dLookUpKeys)
+	imw.innerExec, err = imw.readerBuilder.buildExecutorForIndexJoin(ctx, dLookUpKeys, imw.indexRanges, imw.keyOff2IdxOff, imw.nextColCompareFilters)
+	if err != nil {
+		return err
+	}
+	defer terror.Call(imw.innerExec.Close)
+	_, err = imw.fetchInnerResult(ctx, task)
+	if err != nil {
+		return err
+	}
+	err = imw.doMergeJoin(ctx, task)
+	return err
+}
+
+func (imw *innerMergeWorker) checkChunkIsFull(ctx context.Context, task *lookUpMergeJoinTask, chk **chunk.Chunk) bool {
+	if (*chk).IsFull() {
+		select {
+		case task.results <- *chk:
+		case <-ctx.Done():
+			return true
+		}
+		*chk = chunk.NewChunkWithCapacity(imw.retFieldTypes, imw.maxChunkSize)
+	}
+	return false
+}
+
+func (imw *innerMergeWorker) doMergeJoin(ctx context.Context, task *lookUpMergeJoinTask) (err error) {
+	chk := chunk.NewChunkWithCapacity(imw.retFieldTypes, imw.maxChunkSize)
+	defer func() {
+		if chk.NumRows() > 0 {
+			select {
+			case task.results <- chk:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	if task.innerResult == nil || task.innerResult.NumRows() == 0 {
+		for outerRow := task.outerIter.Begin(); outerRow != task.outerIter.End(); outerRow = task.outerIter.Next() {
+			imw.joiner.onMissMatch(false, outerRow, chk)
+			if imw.checkChunkIsFull(ctx, task, &chk) {
+				return nil
+			}
+		}
+		return
+	}
+
+	task.sameKeyIter = chunk.NewIterator4Slice(task.sameKeyInnerRows)
+	initCmpResult := 1
+	if imw.innerMergeCtx.desc {
+		initCmpResult = -1
+	}
+
+	for outerRow := task.outerIter.Begin(); outerRow != task.outerIter.End(); outerRow = task.outerIter.Next() {
+		if len(task.outerMatch) > 0 && !task.outerMatch[outerRow.Idx()] {
+			continue
+		}
+		cmpResult := initCmpResult
+		if len(task.sameKeyInnerRows) > 0 {
+			cmpResult, err = imw.compare(outerRow, task.sameKeyIter.Begin())
+			if err != nil {
+				return err
+			}
+		}
+		if (cmpResult > 0 && !imw.innerMergeCtx.desc) || (cmpResult < 0 && imw.innerMergeCtx.desc) {
+			err = imw.fetchNextInnerRows(ctx, task, outerRow)
+			if err != nil {
+				return err
+			}
+		}
+
+		hasMatch, hasNull := false, false
+		for task.sameKeyIter.Current() != task.sameKeyIter.End() {
+			matched, isNull, err := imw.joiner.tryToMatchInners(outerRow, task.sameKeyIter, chk)
+			if err != nil {
+				return err
+			}
+
+			hasMatch = hasMatch || matched
+			hasNull = hasNull || isNull
+			if imw.checkChunkIsFull(ctx, task, &chk) {
+				return nil
+			}
+		}
+
+		if !hasMatch {
+			imw.joiner.onMissMatch(hasNull, outerRow, chk)
+		}
+		if imw.checkChunkIsFull(ctx, task, &chk) {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (imw *innerMergeWorker) fetchNextInnerRows(ctx context.Context, task *lookUpMergeJoinTask, key chunk.Row) (err error) {
+	if task.innerResult == nil {
+		return nil
+	}
+	task.sameKeyInnerRows = task.sameKeyInnerRows[:0]
+	defer func() {
```

Comment:
Why put these lines in `defer`?
Can we just put them at the end of this function?

Reply:
No, for safety. We should make sure the `sameKeyIter` is changed when the `sameKeyInnerRows` change.